README.md
unsafe_extension
=====

Library for performing operations on the unmanaged native code and data.

Version: 0.0.24

**Installation**

Go to any application that uses `unsafe_extension` and run

```
$ pub global activate unsafe_extension
```

**Precompiled binaries**  

- ARM Android
- ARM Linux
- X86 Linux
- X86 Windows
- X86_64 Linux
- X86_64 Mac OS X
- X86_64 Windows

Special thanks to [Kenneth Endfinger](https://github.com/kaendfinger) for compiling the binaries:

- ARM Android
- ARM Linux
- X86_64 Mac OS X


CHANGELOG.md
## 0.0.24

- Adapted for use package `sandbox`

## 0.0.21

- Fixed bug in `NativeExtensionBuilder._resolvePackagePath()`

## 0.0.20

- Adapted for use package `package_config`
- Adapted for use package `semaphore`

## 0.0.18

- Update Mac x86_64 Build. Author [Kenneth Endfinger](https://github.com/kaendfinger)

## 0.0.14

- Breaking changes: class `ForeignFunctionInterface` was removed

## 0.0.11

- Initial support of Linux (ARM). Author [Kenneth Endfinger](https://github.com/kaendfinger)

## 0.0.10

- Initial support of Android (ARM). Author [Kenneth Endfinger](https://github.com/kaendfinger)
- Removed support of Mac OS X (X86)

## 0.0.9

- Added the handler `Segmentation Fault` with a post mortem stack trace. Unfortunately it doesn't not work on Windows because setting up the handler  correctly requires a `main` thread of the process

## 0.0.8

- Fixes for Mac OSX

## 0.0.7

- Fixed bug with installation transformer

## 0.0.5

- Added compiled binaries for Mac OS X

## 0.0.4

- Renamed function `call` to `callFunction` in class `ForeignFucntionInterface`

## 0.0.3

- Modified the setup procedure. Installer now finds the latest version of package in cache and setup it.

## 0.0.1

- Initial release



test/test_memory.dart
import 'package:unsafe_extension/unsafe_extension.dart';
import 'package:unittest/unittest.dart';

void main() {
  testLimits();
  testRelativeMemoryAccess();
  print("Done");
}

void testLimits() {
  var limits = <List<int>>[];
  limits.add([-128, 127, 255]);
  limits.add([-32768, 32767, 65535]);
  limits.add([-2147483648, 2147483647, 4294967295]);
  limits.add([-9223372036854775808, 9223372036854775807, 18446744073709551615]);
  var memoryBlock = new MemoryBlock(8);
  var length = 4;
  for (var test = 0; test < 4; test++) {
    var values = limits[test];
    var minSigned = values[0];
    var maxSigned = values[1];
    var minUnsigned = 0;
    var maxUnsigned = values[2];
    // Signed
    var signedValues = <List<int>>[];
    signedValues.add([minSigned, minSigned]);
    signedValues.add([minSigned - 1, maxSigned]);
    signedValues.add([minSigned - 2, maxSigned - 1]);
    signedValues.add([maxSigned, maxSigned]);
    signedValues.add([maxSigned + 1, minSigned]);
    signedValues.add([maxSigned + 2, minSigned + 1]);
    signedValues.add([minUnsigned, minUnsigned]);
    signedValues.add([minUnsigned - 1, -1]);
    signedValues.add([minUnsigned - 2, -2]);
    signedValues.add([maxUnsigned, -1]);
    signedValues.add([maxUnsigned + 1, 0]);
    signedValues.add([maxUnsigned + 2, 1]);
    var length = signedValues.length;
    for (var i = 0; i < length; i++) {
      var input = signedValues[i][0];
      var output = signedValues[i][1];
      String reason;
      int actual;
      switch (test) {
        case 0:
          Unsafe.writeInt8(memoryBlock.address, 0, input);
          actual = Unsafe.readInt8(memoryBlock.address, 0);
          reason = "writeInt8: $input, readInt8: $output";
          break;
        case 1:
          Unsafe.writeInt16(memoryBlock.address, 0, input);
          actual = Unsafe.readInt16(memoryBlock.address, 0);
          reason = "writeInt16: $input, readInt16: $output";
          break;
        case 2:
          Unsafe.writeInt32(memoryBlock.address, 0, input);
          actual = Unsafe.readInt32(memoryBlock.address, 0);
          reason = "writeInt32: $input, readInt32: $output";
          break;
        case 3:
          Unsafe.writeInt64(memoryBlock.address, 0, input);
          actual = Unsafe.readInt64(memoryBlock.address, 0);
          reason = "writeInt64: $input, readInt64: $output";
          break;
      }

      expect(actual, output, reason: reason);
    }

    // Unsigned
    var unsignedValues = <List<int>>[];
    unsignedValues.add([minSigned, -minSigned]);
    unsignedValues.add([minSigned - 1, maxSigned]);
    unsignedValues.add([minSigned - 2, maxSigned - 1]);
    unsignedValues.add([maxSigned, maxSigned]);
    unsignedValues.add([maxSigned + 1, maxSigned + 1]);
    unsignedValues.add([maxSigned + 2, maxSigned + 2]);
    unsignedValues.add([minUnsigned, minUnsigned]);
    unsignedValues.add([minUnsigned - 1, maxUnsigned]);
    unsignedValues.add([minUnsigned - 2, maxUnsigned - 1]);
    unsignedValues.add([maxUnsigned, maxUnsigned]);
    unsignedValues.add([maxUnsigned + 1, 0]);
    unsignedValues.add([maxUnsigned + 2, 1]);
    length = unsignedValues.length;
    for (var i = 0; i < length; i++) {
      var input = unsignedValues[i][0];
      var output = unsignedValues[i][1];
      String reason;
      int actual;
      switch (test) {
        case 0:
          Unsafe.writeUint8(memoryBlock.address, 0, input);
          actual = Unsafe.readUint8(memoryBlock.address, 0);
          reason = "writeUInt8: $input, readUInt8: $output";
          break;
        case 1:
          Unsafe.writeUint16(memoryBlock.address, 0, input);
          actual = Unsafe.readUint16(memoryBlock.address, 0);
          reason = "writeUInt16: $input, readUInt16: $output";
          break;
        case 2:
          Unsafe.writeUint32(memoryBlock.address, 0, input);
          actual = Unsafe.readUint32(memoryBlock.address, 0);
          reason = "writeUInt32: $input, readUInt32: $output";
          break;
        case 3:
          Unsafe.writeUint64(memoryBlock.address, 0, input);
          actual = Unsafe.readUint64(memoryBlock.address, 0);
          reason = "writeUInt64: $input, readUInt64: $output";
          break;
      }

      expect(actual, output, reason: reason);
    }
  }

  // Free memory
  memoryBlock = null;
}

void testRelativeMemoryAccess() {
  var offsets = [-1, 0, 1];
  var count = 3;
  var memoryBlock = new MemoryBlock(count * 8);
  var address = memoryBlock.address + 1;
  for (var i in offsets) {
    var expected = i;
    Unsafe.writeInt8(address + i, 0, i);
    var value = Unsafe.readInt8(address + i, 0);
    expect(value, expected, reason: "read/write int8 $i");
    Unsafe.writeInt16(address + i, 0, i);
    value = Unsafe.readInt16(address + i, 0);
    expect(value, expected, reason: "read/write int16 $i");
    Unsafe.writeInt32(address + i, 0, i);
    value = Unsafe.readInt32(address + i, 0);
    expect(value, expected, reason: "read/write int32 $i");
    Unsafe.writeInt64(address + i, 0, i);
    value = Unsafe.readInt64(address + i, 0);
    expect(value, expected, reason: "read/write int64 $i");
  }

  // Free memory
  memoryBlock = null;
}


test/test_sigsegv.dart
import "package:unsafe_extension/src/unsafe_extension.dart";

void main() {
  Unsafe.readInt32(0, 1);
}


tool/project.dart
import "dart:io";
import "package:build_tools/build_shell.dart";
import "package:build_tools/build_tools.dart";
import "package:build_tools/build_utils.dart";
import "package:file_utils/file_utils.dart";

const String CHANGE_LOG = "tool/change.log";
const String CHANGELOG_MD = "CHANGELOG.md";
const String PUBSPEC_YAML = "pubspec.yaml";
const String README_MD = "README.md";
const String README_MD_IN = "tool/README.md.in";
const String UNSAFE_EXTENSION_INF = "lib/src/unsafe_extension.inf";

void main(List<String> args) {
  // Change directory to root
  FileUtils.chdir("..");

  file(CHANGELOG_MD, [CHANGE_LOG], (Target t, Map args) {
    writeChangelogMd();
  });

  file(README_MD_IN, [], (Target t, Map args) {
    FileUtils.touch([t.name], create: true);
  });

  file(README_MD, [README_MD_IN, PUBSPEC_YAML], (Target t, Map args) {
    var sources = t.sources.toList();
    var template = new File(sources.removeAt(0)).readAsStringSync();
    // Remove "pubspec.yaml"
    sources.removeAt(0);
    for (var filename in sources) {
      var text = new File(filename).readAsStringSync();
      template = template.replaceFirst("{{$filename}}", text);
    }

    template = template.replaceFirst("{{DESCRIPTION}}", getDescription());
    template = template.replaceAll("{{VERSION}}", getVersion());
    //

    new File(t.name).writeAsStringSync(template);
  });

  file(UNSAFE_EXTENSION_INF, [PUBSPEC_YAML], (Target t, Map args) {
    new File(t.name).writeAsStringSync(getVersion());
  });

  target("default", ["git:status"], null, description: "git status");

  target("git:status", [], (Target t, Map args) {
    return exec("git", ["status", "--short"]);
  }, description: "git status --short");

  target("git:add", [], (Target t, Map args) {
    return exec("git", ["add", "--all"]);
  }, description: "git add --all");

  target("git:commit", [CHANGELOG_MD, README_MD, UNSAFE_EXTENSION_INF, "git:add"], (Target t, Map args) {
    var message = args["m"];
    if (message == null || message.isEmpty) {
      print("Please, specify the `commit` message with --m option");
      return -1;
    }

    return exec("git", ["commit", "-m", message]);
  }, description: "git commit, --m \"message\"");

  after(["git:commit"], (Target t, Map args) {
    var version = incrementVersion(getVersion());
    print("Change the project version to '$version' (Y/N)?");
    if (stdin.readLineSync().toLowerCase().startsWith("y")) {
      updateVersion(version);
      print("Version switched to $version");
    }
  });

  target("git:push", [], (Target t, Map args) {
    // TODO: The `exec git push` does not show the prompt on Windows.
    // But on Posix everything works as expected.
    // Problem in the "git" or in the Dart VM "stdxxx" streams.
    return exec("git", ["push", "origin", "master"]);
  }, description: "git push origin master");

  target("log:changes", [], (Target t, Map args) {
    var message = args["m"];
    if (message == null || message.isEmpty) {
      print("Please, specify the `message` with --m option");
      return -1;
    }

    logChanges(message);
  }, description: "log changes, --m message", reusable: true);

  target("prj:changelog", [CHANGELOG_MD], null, description: "generate '$CHANGELOG_MD'", reusable: true);

  target("prj:readme", [README_MD], null, description: "generate '$README_MD'", reusable: true);

  target("prj:version", [], (Target t, Map args) {
    print("Version: ${getVersion()}");
  }, description: "display version", reusable: true);

  new BuildShell().run(args).then((exitCode) => exit(exitCode));
}

String getDescription() {
  var file = new File(PUBSPEC_YAML);
  var lines = file.readAsLinesSync();
  var description = "";
  for (var line in lines) {
    if (line.startsWith("description")) {
      var index = line.indexOf(":");
      if (index != -1 && line.length > index + 1) {
        description = line.substring(index + 1).trim();
      }
    }
  }

  return description;
}

String getVersion() {
  var file = new File(PUBSPEC_YAML);
  var lines = file.readAsLinesSync();
  var version = "0.0.1";
  for (var line in lines) {
    if (line.startsWith("version")) {
      var index = line.indexOf(":");
      if (index != -1 && line.length > index + 1) {
        version = line.substring(index + 1).trim();
      }
    }
  }

  return version;
}

String incrementVersion(String version) {
  var parts = version.split(".");
  if (parts.length < 3) {
    return version;
  }

  var patch = int.parse(parts[2], onError: (x) => null);
  if (patch == null) {
    return version;
  }

  parts[2] = ++patch;
  parts.length = 3;
  return parts.join(".");
}

void logChanges(String message) {
  if (message == null || message.isEmpty) {
    return;
  }

  FileUtils.touch([CHANGE_LOG], create: true);
  var file = new File(CHANGE_LOG);
  var length = file.lengthSync();
  var fp = file.openSync(mode: FileMode.APPEND);
  var sb = new StringBuffer();
  if (length != 0) {
    sb.writeln();
  }

  sb.write("${getVersion()} $message");
  fp.writeStringSync(sb.toString());
  fp.closeSync();
}

void updateVersion(String version) {
  var file = new File(PUBSPEC_YAML);
  var found = false;
  var lines = file.readAsLinesSync();
  var sb = new StringBuffer();
  for (var line in lines) {
    if (line.startsWith("version")) {
      found = true;
      break;
    }
  }

  if (!found) {
    var pos = lines.length == 0 ? 0 : 1;
    lines.insert(pos, "version: $version");
  }

  for (var line in lines) {
    if (line.startsWith("version")) {
      sb.writeln("version: $version");
    } else {
      sb.writeln(line);
    }
  }

  var string = sb.toString();
  file.writeAsStringSync(string);
}

void writeChangelogMd() {
  FileUtils.touch([CHANGELOG_MD], create: true);
  var log = new File(CHANGE_LOG);
  if (!log.existsSync()) {
    return;
  }

  var lines = log.readAsLinesSync();
  lines = lines.reversed.toList();
  var versions = <String, List<String>>{};
  for (var line in lines) {
    var index = line.indexOf(" ");
    if (index != -1) {
      var version = line.substring(0, index);
      var message = line.substring(index + 1).trimLeft();
      var messages = versions[version];
      if (messages == null) {
        messages = <String>[];
        versions[version] = messages;
      }

      messages.add(message);
    }
  }

  var sb = new StringBuffer();
  for (var version in versions.keys) {
    sb.writeln("## ${version}");
    sb.writeln();
    var messages = versions[version];
    messages.sort((a, b) => a.compareTo(b));
    for (var message in messages) {
      sb.writeln("- $message");
    }

    sb.writeln();
  }

  var md = new File(CHANGELOG_MD);
  md.writeAsStringSync(sb.toString());
}


lib/src/unsafe_extension.dart
library unsafe_extension.unsafe_extension;

import 'dart-ext:unsafe_extension';

/**
 * Memory block. Allocates memory when creating, frees memory when destroying.
 */
class MemoryBlock {
  /**
   * Size of memory block in bytes.
   */
  final int size;

  int _address;

  MemoryBlock(this.size) {
    if (size == null || size <= 0) {
      throw new ArgumentError.notNull("size");
    }

    if (size <= 0) {
      throw new ArgumentError.value(size, "size");
    }

    _address = Unsafe.memoryAllocate(size);
    if (_address == 0) {
      throw new StateError("Cannot allocate memory");
    }

    Unsafe.memoryPeer(this, _address, size);
  }

  /**
   * Returns the address of memory block.
   */
  int get address => _address;
}

/**
 * Unsafe operations helper.
 */
class Unsafe {
  /**
   * Indicates when the endianness is little endian.
   */
  static final bool isLittleEndian = _isLittleEndian();

  /**
   * Size of the "pointer" type.
   */
  static final int sizeOfPointer = _getSizeOfPointer();

  /**
   * Page size of physical memory.
   */
  static final int pageSize = _getPageSize();

  /**
   * Frees the dynamic library.
   *
   * Parameters:
   *   [int] handle
   *   Handle of dynamic library.
   */
  static int libraryFree(int handle) {
    if (handle == null) {
      throw new ArgumentError.notNull("handle");
    }

    return _libraryFree(handle);
  }

  /**
   * Loads the dynamic library.
   *
   * Parameters:
   *   [String] filename
   *   File name of the dynamic library.
   */
  static int libraryLoad(String filename) {
    if (filename == null) {
      throw new ArgumentError.notNull("filename");
    }

    if (filename.isEmpty) {
      throw new ArgumentError("File name should not be empty");
    }

    return _libraryLoad(filename);
  }

  /**
   * Obtains the address of a symbol from a dynamic library.
   *
   * Parameters:
   *   [int] handle
   *   Handle of dynamic library.
   *
   *   [String] symbol
   *   Symbol to obtain an address.
   */
  static int librarySymbol(int handle, String symbol) {
    if (handle == null) {
      throw new ArgumentError.notNull("handle");
    }

    if (handle == 0) {
      throw new ArgumentError.value(handle , "handle");
    }

    if (symbol == null) {
      throw new ArgumentError.notNull("symbol");
    }

    if (symbol.isEmpty) {
      throw new ArgumentError("Symbol should not be empty");
    }

    return _librarySymbol(handle, symbol);
  }

  static int memoryAllocate(int size) {
    if (size == null) {
      throw new ArgumentError.notNull("size");
    }

    if (size < 0) {
      throw new ArgumentError.value(size, "size");
    }

    return _memoryAllocate(size);
  }

  /**
   * Copies the values of num bytes from the source to the destionation.
   * Does not allows the destination and source to overlap.
   *
   * Parameters:
   *   [int] dest
   *   Address of destination.
   *
   *   [int] src
   *   Address of source.
   *
   *   [int] num
   *   Numbers of bytes to copy.
   */
  static void memoryCopy(int dest, int src, int num) {
    if (dest == null) {
      throw new ArgumentError.notNull("dest");
    }

    if (src == null) {
      throw new ArgumentError.notNull("src");
    }

    if (num == null) {
      throw new ArgumentError.notNull("num");
    }

    if (num < 0) {
      throw new ArgumentError.value(num, "num");
    }

    _memoryCopy(dest, src, num);
  }

  /**
   * Frees the allocated memory.
   *
   * Parameters:
   *   [int] handle
   *   Address of allocated memory.
   */
  static void memoryFree(int handle) {
    if (handle == null) {
      throw new ArgumentError.notNull("handle");
    }

    _memoryFree(handle);
  }

  /**
   * Copies the values of num bytes from the source to the destionation.
   * Allows the destination and source to overlap.
   *
   * Parameters:
   *   [int] dest
   *   Address of destination.
   *
   *   [int] src
   *   Address of source.
   *
   *   [int] num
   *   Numbers of bytes to copy.
   */
  static void memoryMove(int dest, int src, int num) {
    if (dest == null) {
      throw new ArgumentError.notNull("dest");
    }

    if (src == null) {
      throw new ArgumentError.notNull("src");
    }

    if (num == null) {
      throw new ArgumentError.notNull("num");
    }

    _memoryMove(dest, src, num);
  }

  /**
   * Registers the peer with the specified object and returns the weak persistence handle.
   *
   * Parameters:
   *   [Object] object
   *
   *   [int] peer
   *   Address of peer.
   *
   *   [int] size
   *   Size of the peer.
   */
  static int memoryPeer(Object object, int peer, int size) {
    if (object == null || object is bool || object is num || object is String) {
      throw new ArgumentError.value(object, "object");
    }

    if (peer == null) {
      throw new ArgumentError.notNull("peer");
    }

    if (size == null) {
      throw new ArgumentError.notNull("size");
    }

    if (size < 0) {
      throw new ArgumentError.value(size, "size");
    }

    return _peerRegister(object, peer, size);
  }

  /**
   * Sets the first num bytes of the memory the specified value.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to set.
   *
   *   [int] num
   *   Number of bytes to be set to the value.
   */
  static void memorySet(int base, int offset, int value, int num) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    if (num == null) {
      throw new ArgumentError.notNull("num");
    }

    if (num < 0) {
      throw new ArgumentError.value(num, "num");
    }

    value &= 0xff;
    if (value < 0) {
      value = 255 + value + 1;
    }

    _memorySet(base, offset, value, num);
  }

  /**
   * Reads the 32-bit float value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [double] value
   *   Value to read.
   */
  static double readFloat32(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readFloat32(base, offset);
  }

  /**
   * Reads the 64-bit float value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [double] value
   *   Value to read.
   */
  static double readFloat64(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readFloat64(base, offset);
  }

  /**
   * Reads the 16-bit signed int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readInt16(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readInt16(base, offset);
  }

  /**
   * Reads the 32-bit signed int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readInt32(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readInt32(base, offset);
  }

  /**
   * Reads the 64-bit signed int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readInt64(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readInt64(base, offset);
  }

  /**
   * Reads the 8-bit signed int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readInt8(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readInt8(base, offset);
  }

  /**
   * Reads the integer pointer value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readIntPtr(int base, int offset) {
    switch (Unsafe.sizeOfPointer) {
      case 4:
        return readInt32(base, offset);
      case 8:
        return readInt64(base, offset);
      default:
        throw ('Illegal size (${Unsafe.sizeOfPointer}) of IntPtr');
    }
  }

  /**
   * Reads the 16-bit unsigned int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readUint16(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readUint16(base, offset);
  }

  /**
   * Reads the 32-bit unsigned int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readUint32(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readUint32(base, offset);
  }

  /**
   * Reads the 64-bit unsigned int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readUint64(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    var value = _readUint64(base, offset);
    // Dart API limitation does not allow to create an unsigned 64-bit integers
    if (value < 0) {
      value = 18446744073709551615 + value + 1;
    }

    return value;
  }

  /**
   * Reads the 8-bit unsigned int value from memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to read.
   */
  static int readUint8(int base, int offset) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    return _readUint8(base, offset);
  }

  /**
   * Writes the 32-bit float value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [double] value
   *   Value to write.
   */
  static void writeFloat32(int base, int offset, double value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    _writeFloat32(base, offset, value);
  }

  /**
   * Writes the 64-bit float value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [double] value
   *   Value to write.
   */
  static void writeFloat64(int base, int offset, double value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    _writeFloat64(base, offset, value);
  }

  /**
   * Writes the 16-bit signed integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeInt16(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffff;
    _writeInt16(base, offset, value);
  }

  /**
   * Writes the 32-bit signed integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeInt32(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffffffff;
    _writeInt32(base, offset, value);
  }

  /**
   * Writes the 64-bit signed integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeInt64(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffffffffffffffff;
    if (value > 9223372036854775807) {
      value = value - 0x10000000000000000;
    }

    _writeInt64(base, offset, value);
  }

  /**
   * Writes the 8-bit signed integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeInt8(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xff;
    _writeInt8(base, offset, value);
  }

  static void writeIntPtr(int base, int offset, int ptr) {
    if (ptr == null) {
      throw new ArgumentError.notNull("ptr");
    }

    switch (Unsafe.sizeOfPointer) {
      case 4:
        writeInt32(base, offset, ptr);
        break;
      case 8:
        writeInt64(base, offset, ptr);
        break;
      default:
        throw ('Illegal size (${Unsafe.sizeOfPointer}) of IntPtr');
    }
  }

  /**
   * Writes the 16-bit unsigned integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeUint16(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffff;
    _writeUint16(base, offset, value);
  }

  /**
   * Writes the 32-bit unsigned integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeUint32(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffffffff;
    _writeUint32(base, offset, value);
  }

  /**
   * Writes the 64-bit unsigned integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeUint64(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xffffffffffffffff;
    _writeUint64(base, offset, value);
  }

  /**
   * Writes the 8-bit unsigned integer value in memory.
   *
   * Parameters:
   *   [int] base
   *   Base of memory address.
   *
   *   [int] offset
   *   Offset of memory address.
   *
   *   [int] value
   *   Value to write.
   */
  static void writeUint8(int base, int offset, int value) {
    if (base == null) {
      throw new ArgumentError.notNull("base");
    }

    if (offset == null) {
      throw new ArgumentError.notNull("offset");
    }

    if (value == null) {
      throw new ArgumentError.notNull("value");
    }

    value &= 0xff;
    _writeUint8(base, offset, value);
  }

  static int _getPageSize() native 'Unsafe_GetPageSize';

  static int _getSizeOfPointer() native 'Unsafe_GetSizeOfPointer';

  static bool _isLittleEndian() native 'Unsafe_IsLittleEndian';

  static int _libraryFree(int handle) native 'Unsafe_LibraryFree';

  static int _libraryLoad(String name) native 'Unsafe_LibraryLoad';

  static int _librarySymbol(int handle, String symbol) native 'Unsafe_LibrarySymbol';

  static int _memoryAllocate(int size) native 'Unsafe_MemoryAllocate';

  static void _memoryCopy(int dest, int src, int num) native 'Unsafe_MemoryCopy';

  static void _memoryFree(int handle) native 'Unsafe_MemoryFree';

  static void _memoryMove(int dest, int src, int num) native 'Unsafe_MemoryMove';

  static void _memorySet(int base, int offset, int value, int num) native 'Unsafe_MemorySet';

  static int _peerRegister(Object object, int peer, int size) native 'Unsafe_PeerRegister';

  static double _readFloat32(int base, int offset) native 'Unsafe_ReadFloat32';

  static double _readFloat64(int base, int offset) native 'Unsafe_ReadFloat64';

  static int _readInt16(int base, int offset) native 'Unsafe_ReadInt16';

  static int _readInt32(int base, int offset) native 'Unsafe_ReadInt32';

  static int _readInt64(int base, int offset) native 'Unsafe_ReadInt64';

  static int _readInt8(int base, int offset) native 'Unsafe_ReadInt8';

  static int _readUint16(int base, int offset) native 'Unsafe_ReadUInt16';

  static int _readUint32(int base, int offset) native 'Unsafe_ReadUInt32';

  static int _readUint64(int base, int offset) native 'Unsafe_ReadUInt64';

  static int _readUint8(int base, int offset) native 'Unsafe_ReadUInt8';

  static void _writeFloat32(int base, int offset, double value) native 'Unsafe_WriteFloat32';

  static void _writeFloat64(int base, int offset, double value) native 'Unsafe_WriteFloat64';

  static void _writeInt16(int base, int offset, int value) native 'Unsafe_WriteInt16';

  static void _writeInt32(int base, int offset, int value) native 'Unsafe_WriteInt32';

  static void _writeInt64(int base, int offset, int value) native 'Unsafe_WriteInt64';

  static void _writeInt8(int base, int offset, int value) native 'Unsafe_WriteInt8';

  static void _writeUint16(int base, int offset, int value) native 'Unsafe_WriteUInt16';

  static void _writeUint32(int base, int offset, int value) native 'Unsafe_WriteUInt32';

  static void _writeUint64(int base, int offset, int value) native 'Unsafe_WriteUInt64';

  static void _writeUint8(int base, int offset, int value) native 'Unsafe_WriteUInt8';
}
